### DISCUSSION TITLE
bundle.js não encontrado em projeto MVC em servidor

### DISCUSSION THREAD

**Discussion by rimancete** [frontendbr/forum#1922] [Perguntas e Respostas]
Olá pessoal.
O seguinte projeto [https://github.com/rimancete/js_Avancado/tree/master/node/agenda](https://github.com/rimancete/js_Avancado/tree/master/node/agenda) não está carregando o bundle.js no servidor, mas localmente funciona perfeitamente.

Estou estudando dev a alguns meses, então perdoai se a pergunta for muito básica rs, mas estou a 3 dias tentando resolver isso... criei, recriei alguns servidores na AWS EC2 e Google Cloud Plataform. Nos servers estou utilizando nginx como proxy reverso.
O sistema está disponível em [http://agenda.lojardente.com/](http://agenda.lojardente.com/).

Pesquisei no stack, aqui mesmo e não encontro um norte para resolver isso!

O projeto é somente para estudo, posso compartilhar aqui o que for necessário para entendimento.

Muito obrigado pela ajuda!

**Comment by rimancete** [comment]
**_package.json_**

```
{
  "name": "agenda",
  "version": "1.0.2",
  "description": "",
  "main": "server.js",
  "scripts": {
    "start": "nodemon server.js --ignore public --ignore frontend",
    "dev": "webpack -w"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "bcryptjs": "^2.4.3",
    "bootstrap": "^4.5.3",
    "connect-flash": "^0.1.1",
    "connect-mongo": "^3.2.0",
    "core-js": "^3.6.5",
    "css-loader": "^5.0.0",
    "csurf": "^1.11.0",
    "dotenv": "^8.2.0",
    "ejs": "^3.1.5",
    "express": "^4.17.1",
    "express-session": "^1.17.1",
    "helmet": "^4.2.0",
    "jquery": "^3.5.1",
    "mongoose": "^5.10.12",
    "popper.js": "^1.16.1",
    "regenerator-runtime": "^0.13.7",
    "style-loader": "^2.0.0",
    "validator": "^13.1.17"
  },
  "devDependencies": {
    "nodemon": "^2.0.6",
    "@babel/cli": "^7.12.1",
    "@babel/core": "^7.12.3",
    "@babel/preset-env": "^7.12.1",
    "babel-loader": "^8.1.0",
    "webpack": "^4.44.2",
    "webpack-cli": "^3.3.12"
  }
}
```
**_webpack.config.js_**

```
const path = require('path'); // CommonJS

module.exports = {
  mode: 'production',
  entry: './frontend/main.js',
  output: {
    publicPath: "/public/",
    path: path.resolve(__dirname, 'public', 'assets', 'js'),
    filename: 'bundle.js'
  },
  module: {
    rules: [{
      exclude: /node_modules/,
      test: /\.js$/,
      use: {
        loader: 'babel-loader',
        options: {
          presets: ['@babel/env']
        }
      }
    }, {
      test: /\.css$/,
      use: ['style-loader', 'css-loader']
    }]
  },
  devtool: 'source-map'
};
```
**_package.json_**

```
  {
  "name": "agenda",
  "version": "1.0.2",
  "description": "",
  "main": "server.js",
  "scripts": {
    "start": "nodemon server.js --ignore public --ignore frontend",
    "dev": "webpack -w"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "bcryptjs": "^2.4.3",
    "bootstrap": "^4.5.3",
    "connect-flash": "^0.1.1",
    "connect-mongo": "^3.2.0",
    "core-js": "^3.6.5",
    "css-loader": "^5.0.0",
    "csurf": "^1.11.0",
    "dotenv": "^8.2.0",
    "ejs": "^3.1.5",
    "express": "^4.17.1",
    "express-session": "^1.17.1",
    "helmet": "^4.2.0",
    "jquery": "^3.5.1",
    "mongoose": "^5.10.12",
    "popper.js": "^1.16.1",
    "regenerator-runtime": "^0.13.7",
    "style-loader": "^2.0.0",
    "validator": "^13.1.17"
  },
  "devDependencies": {
    "nodemon": "^2.0.6",
    "@babel/cli": "^7.12.1",
    "@babel/core": "^7.12.3",
    "@babel/preset-env": "^7.12.1",
    "babel-loader": "^8.1.0",
    "webpack": "^4.44.2",
    "webpack-cli": "^3.3.12"
  }
}
```
**_footer.ejs_**

```
<div class="col-lg-2"></div>
</div>
</section>

<script src="/assets/js/bundle.js"></script>
</body>

</html>
```
**_estrutura de pasta_**
![agenda-folder](https://user-images.githubusercontent.com/47704038/99325972-2354f180-2856-11eb-9fc7-04189a4e0a8d.png)

**Comment by renedocket** [comment]
@rimancete Cara eu acredito que você precise adicionar algo como `filename: './js/[name].js'`, para ele entender que são vários arquivos e gerar um bundle.js pra cada se é isso que eu entendi o que você quer fazer.
E também tenta criar um script de build algo como:
 ```
 "scripts": {
    "build": "webpack -p --config ./webpack.config.js --display-error-details",
  },
```

Vê se ajuda. :)

**Comment by rimancete** [comment]
@renedocket , na verdade eu só uso o bundle.js nesse projeto mesmo além o map para tratar erros, mas fiz o que sugeriu e o resultado é o mesmo. Localmente tudo funciona perfeitamente. No servidor ele não é carregado.
Inclui o script de build no package.json e ele mostra alerta sobre o tamanho do meu entry(acredito não tenha relação com o foco desse tópico, então ficaria pra um próximo post tentar solucioná-lo rs) O objetivo agora é fazer meu sistema conseguir carregar o bundle...

**_log do webpack_**

```
> webpack -p --config ./webpack.config.js --display-error-details

Hash: 44179dec395f27b4ac3f
Version: webpack 4.44.2
Time: 5588ms
Built at: 11/17/2020 12:11:44 PM
        Asset      Size  Chunks                          Chunk Names
    bundle.js  1.29 MiB       0  [emitted]        [big]  main
bundle.js.map   562 KiB       0  [emitted] [dev]         main
Entrypoint main [big] = bundle.js bundle.js.map
[121] ./node_modules/css-loader/dist/cjs.js!./frontend/assets/css/style.css 3.66 KiB {0} [built]
[122] (webpack)/buildin/global.js 472 bytes {0} [built]
[479] ./frontend/main.js + 5 modules 16.8 KiB {0} [built]
      | ./frontend/main.js 853 bytes [built]
      | ./frontend/assets/css/style.css 338 bytes [built]
      | ./frontend/modules/Login.js 4.32 KiB [built]
      | ./frontend/modules/Signup.js 6.17 KiB [built]
      | ./frontend/modules/Contato.js 4.84 KiB [built]
      |     + 1 hidden module
    + 477 hidden modules

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  bundle.js (1.29 MiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  main (1.29 MiB)
      bundle.js


WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/
```
Até o momento o nginx não gerou log de erro...

**Comment by renedocket** [comment]
> @renedocket , na verdade eu só uso o bundle.js nesse projeto mesmo além o map para tratar erros, mas fiz o que sugeriu e o resultado é o mesmo. Localmente tudo funciona perfeitamente. No servidor ele não é carregado.
> Inclui o script de build no package.json e ele mostra alerta sobre o tamanho do meu entry(acredito não tenha relação com o foco desse tópico, então ficaria pra um próximo post tentar solucioná-lo rs) O objetivo agora é fazer meu sistema conseguir carregar o bundle...
> 
> **_log do webpack_**
> 
> ```
> > webpack -p --config ./webpack.config.js --display-error-details
> 
> Hash: 44179dec395f27b4ac3f
> Version: webpack 4.44.2
> Time: 5588ms
> Built at: 11/17/2020 12:11:44 PM
>         Asset      Size  Chunks                          Chunk Names
>     bundle.js  1.29 MiB       0  [emitted]        [big]  main
> bundle.js.map   562 KiB       0  [emitted] [dev]         main
> Entrypoint main [big] = bundle.js bundle.js.map
> [121] ./node_modules/css-loader/dist/cjs.js!./frontend/assets/css/style.css 3.66 KiB {0} [built]
> [122] (webpack)/buildin/global.js 472 bytes {0} [built]
> [479] ./frontend/main.js + 5 modules 16.8 KiB {0} [built]
>       | ./frontend/main.js 853 bytes [built]
>       | ./frontend/assets/css/style.css 338 bytes [built]
>       | ./frontend/modules/Login.js 4.32 KiB [built]
>       | ./frontend/modules/Signup.js 6.17 KiB [built]
>       | ./frontend/modules/Contato.js 4.84 KiB [built]
>       |     + 1 hidden module
>     + 477 hidden modules
> 
> WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
> This can impact web performance.
> Assets: 
>   bundle.js (1.29 MiB)
> 
> WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
> Entrypoints:
>   main (1.29 MiB)
>       bundle.js
> 
> 
> WARNING in webpack performance recommendations: 
> You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
> For more info visit https://webpack.js.org/guides/code-splitting/
> ```
> 
> Até o momento o nginx não gerou log de erro...

Saquei, pelos warnings o seu bundle.js está bem grande, tenta dividir o seu main.js em outros arquivos algo do tipo
```
module.exports = {
        context: path.resolve(__dirname, 'front-end'),
        mode: 'production',
	entry: {
		main: ['pasta/caminho_arquivo.js'],
                index:['/pasta/caminho']
	}
```
você pode usar a o context, pra sempre pegar os scripts dentro dessa pasta. Acredito que na hora de empacotar o projeto pra prod, você precise especificar qual pasta quer você manda pra prod saca? Não sei ao certo se é isso, mas pelo que vejo talvez essa pasta não esta sendo lida. :thinking:

**Comment by rimancete** [comment]
@renedocket Meu bundle está enorme hehehe, eu vou pesquisar sobre essa distribuição de js...Muito obrigado pela dica :)
Quando tentei a sua sugestão de alterar a chave _filename_ consegui fazer ele gerar novos bundles no mesmo diretório do bundle, ou seja, a forma de especificar o caminho estaria certa, se não eu não chegaria ali, correto?

De qualquer forma também acredito que seja algo na aplicação. Recriei esse servidor umas três vezes pra ver se eu tinha configurado algo errado nele e aparentemente esse problema não é com o server, já que tenho tbm um conteiner mariadb e está 100% acessível. 
Enfim, o que pensei eu fiz, estou sem idéia do que tentar pra resolver :(
